akka-http-core/src/main/scala/akka/http/impl/engine/server/ServerTerminator.scala

```diff
@@ -0,0 +1,294 @@
+  // terminators such that they can initiate the termination (draining, failing) of their respective connections.
+
+  private val terminators = new AtomicReference[MasterServerTerminator.State](MasterServerTerminator.AliveConnectionTerminators(Set.empty))
+  private val termination = Promise[HttpTerminated]()
+
+  /**
+   * Registers an per-connection terminator.
+   * Once the master terminator gets the termination signal,
+   * it will delegate the signal to all registered terminators (existing connections).
+   *
+   * @return true if registered successfully and not terminating, false if termination in-flight
+   *         (and the terminators `terminate()` will be invoked in that case automatically)
+   */
+  @tailrec def registerConnection(terminator: ServerTerminator)(implicit ec: ExecutionContext): Boolean = {
+    terminators.get() match {
+      case v @ AliveConnectionTerminators(ts) ⇒
+        terminators.compareAndSet(v, v.copy(ts = ts + terminator)) ||
+          registerConnection(terminator) // retry
+
+      case Terminating(deadline) ⇒
+        terminator.terminate(deadline.timeLeft)(ec)
+        false // termination is in progress already, we did not register but immediately issue termination
+    }
+  }
+
+  // If a connection attempts to register once termination has started, it will immediately be rejected (though
+  // since termination also implies unbinding such new connections should not really happen).
+  def terminate(timeout: FiniteDuration)(implicit ex: ExecutionContext): Future[HttpTerminated] = {
+    terminators.get() match {
+      case v @ AliveConnectionTerminators(ts) if ts.isEmpty ⇒
+        if (terminators.compareAndSet(v, Terminating(timeout.fromNow))) {
+          // no connections exist to be terminated, and if some arrive after us, they will see the Terminating and terminate
+          termination.success(HttpServerTerminated)
+          termination.future
+        } else terminate(timeout)(ex) // retry
+
+      case v @ AliveConnectionTerminators(ts) ⇒
+        if (terminators.compareAndSet(v, Terminating(timeout.fromNow))) {
+          // cause the termination for all connections
+          val connectionsTerminated = Future.sequence(ts.map { _.terminate(timeout) })
+          val serverTerminated = connectionsTerminated.map(_ ⇒ HttpServerTerminated)
+          termination.completeWith(serverTerminated)
+          termination.future
+        } else terminate(timeout)(ex) // retry
+
+      case Terminating(existingDeadline) ⇒
+        log.warning(s"Issued terminate($timeout) while termination is in progress already (with deadline: time left: ${PrettyDuration.format(existingDeadline.timeLeft)}")
+        termination.future
+    }
+  }
+}
+
+// TODO message better?
+final class ServerTerminationDeadlineReached() extends RuntimeException("Server termination deadline reached, shutting down all connections and terminating server...")
```

Comment:
message does not really matter too much i think... since we handle it internally only, will remove todo